### PR TITLE
Fix parameter detection for expression arguments

### LINKED ISSUE #2537: Unable to cast "System.Linq.Expressions.UnaryExpression" to "System.Linq.Expressions.LambdaExpression".
Hello.
After update from 5.2.7 to 5.3.* we have some issue with custom QueryProvider. I make test application - 
https://github.com/emejibka/nhibernate-specification-subquery-bug

```C#
System.NotSupportedException : Could not parse expression 'value(NHibernate.Linq.NhQueryable`1[TestApp.Client]).Where(Convert(value(TestApp.Specification`1[TestApp.Client])))': The given arguments did not match the expected arguments: Невозможно преобразовать объект типа "System.Linq.Expressions.UnaryExpression" к типу "System.Linq.Expressions.LambdaExpression".
   в Remotion.Linq.Parsing.Structure.MethodCallExpressionParser.CreateExpressionNode(Type nodeType, MethodCallExpressionParseInfo parseInfo, Object[] additionalConstructorParameters)
   в Remotion.Linq.Parsing.Structure.MethodCallExpressionParser.Parse(String associatedIdentifier, IExpressionNode source, IEnumerable`1 arguments, MethodCallExpression expressionToParse)
   в Remotion.Linq.Parsing.Structure.ExpressionTreeParser.ParseMethodCallExpression(MethodCallExpression methodCallExpression, String associatedIdentifier)
   в Remotion.Linq.Parsing.Structure.ExpressionTreeParser.ParseMethodCallExpression(MethodCallExpression methodCallExpression, String associatedIdentifier)
   в Remotion.Linq.Parsing.Structure.ExpressionTreeParser.ParseTree(Expression expressionTree)
   в Remotion.Linq.Parsing.Structure.QueryParser.GetParsedQuery(Expression expressionTreeRoot)
   в Remotion.Linq.Parsing.ExpressionVisitors.SubQueryFindingExpressionVisitor.Visit(Expression expression)
   в System.Linq.Expressions.ExpressionVisitor.VisitLambda[T](Expression`1 node)
   в Remotion.Linq.Parsing.ExpressionVisitors.SubQueryFindingExpressionVisitor.Visit(Expression expression)
   в Remotion.Linq.Parsing.Structure.MethodCallExpressionParser.ProcessArgumentExpression(Expression argumentExpression)
   в System.Linq.Enumerable.WhereSelectEnumerableIterator`2.MoveNext()
   в System.Linq.Buffer`1..ctor(IEnumerable`1 source)
   в System.Linq.Enumerable.ToArray[TSource](IEnumerable`1 source)
   в Remotion.Linq.Parsing.Structure.MethodCallExpressionParser.Parse(String associatedIdentifier, IExpressionNode source, IEnumerable`1 arguments, MethodCallExpression expressionToParse)
   в Remotion.Linq.Parsing.Structure.ExpressionTreeParser.ParseMethodCallExpression(MethodCallExpression methodCallExpression, String associatedIdentifier)
   в Remotion.Linq.Parsing.Structure.ExpressionTreeParser.ParseTree(Expression expressionTree)
   в Remotion.Linq.Parsing.Structure.QueryParser.GetParsedQuery(Expression expressionTreeRoot)
   в NHibernate.Linq.NhLinqExpression.Translate(ISessionFactoryImplementor sessionFactory, Boolean filter)
   в NHibernate.Hql.Ast.ANTLR.ASTQueryTranslatorFactory.CreateQueryTranslators(IQueryExpression queryExpression, String collectionRole, Boolean shallow, IDictionary`2 filters, ISessionFactoryImplementor factory)
   в NHibernate.Engine.Query.QueryPlanCache.GetHQLQueryPlan(IQueryExpression queryExpression, Boolean shallow, IDictionary`2 enabledFilters)
   в NHibernate.Impl.AbstractSessionImpl.GetHQLQueryPlan(IQueryExpression queryExpression, Boolean shallow)
   в NHibernate.Impl.AbstractSessionImpl.CreateQuery(IQueryExpression queryExpression)
   в NHibernate.Linq.DefaultQueryProvider.PrepareQuery(Expression expression, IQuery& query)
   в TestApp.CustomQueryProvider.PrepareQuery(Expression expression, IQuery& query) в C:\temp\nhibernate-specification-subquery-bug\CustomQueryProvider.cs:строка 17
   в NHibernate.Linq.DefaultQueryProvider.ExecuteList[TResult](Expression expression)
   в NHibernate.Linq.NhQueryable`1.System.Collections.Generic.IEnumerable<T>.GetEnumerator()
   в System.Collections.Generic.List`1..ctor(IEnumerable`1 collection)
   в System.Linq.Enumerable.ToList[TSource](IEnumerable`1 source)
   в TestApp.TestClass.test() в C:\temp\nhibernate-specification-subquery-bug\TestClass.cs:строка 48
```

### PR DESCRIPTION
**Pull request by maca88**
Fix for a regression made by #2335.

Fixes #2537

### CODE REVIEW

File: src/NHibernate/Linq/Visitors/NhPartialEvaluatingExpressionVisitor.cs [hunk 1]
```diff
@@ -157,7 +157,9 @@ private Expression EvaluateSubtree(Expression subtree)
 
 		private bool ContainsVariable(Expression expression)
 		{
-			if (!(expression is UnaryExpression unaryExpression))
+			if (!(expression is UnaryExpression unaryExpression) ||
+				// Avoid detecting expression variables as parameters
+				typeof(Expression).IsAssignableFrom(expression.Type.UnwrapIfNullable()))
```
Comment (bahusoid) [inline src/NHibernate/Linq/Visitors/NhPartialEvaluatingExpressionVisitor.cs:162]: `UnwrapIfNullable` looks unnecessary here. Or am I wrong?

Comment (hazzik) [reply src/NHibernate/Linq/Visitors/NhPartialEvaluatingExpressionVisitor.cs:162]: You're correct. `Expression` is a reference type.

File: src/NHibernate/Linq/Visitors/NhPartialEvaluatingExpressionVisitor.cs [hunk 2]
```diff
@@ -157,7 +157,9 @@ private Expression EvaluateSubtree(Expression subtree)
 
 		private bool ContainsVariable(Expression expression)
```
Comment (hazzik) [inline src/NHibernate/Linq/Visitors/NhPartialEvaluatingExpressionVisitor.cs:158]: @maca88 can this method be moved to `NhEvaluatableExpressionFilter`?

Comment (maca88) [reply src/NHibernate/Linq/Visitors/NhPartialEvaluatingExpressionVisitor.cs:158]: Unfortunately not, it was placed here due to #2458, to evaluate expression like `string.Fromat("{0}", variable)`.

File: src/NHibernate/Linq/Visitors/NhPartialEvaluatingExpressionVisitor.cs (hunk 1 above)
Comment (maca88) [reply src/NHibernate/Linq/Visitors/NhPartialEvaluatingExpressionVisitor.cs]: Removed.

**Review by bahusoid — APPROVED** [review]

**Pull request merged**